File: tools/r4-softmax-trainer/src/r4_softmax_trainer/text_clause_adapter/adapter.py
"""Deterministic raw-byte entry to the frozen five-clause reader (#1094).

Only lexical IDs, lengths and raw-byte provenance are produced. The local
grammar recognizer returns booleans; it does not reconstruct a semantic world,
select a role, compute an answer or call the model. Artifact verification and
model execution belong to the wrapper, which must verify startup before
calling ``segment_request``.
"""

from __future__ import annotations

import hashlib
import json
import struct
from collections.abc import Sequence
from pathlib import Path
from types import MappingProxyType
from typing import NamedTuple


POLICY_BYTES = Path(__file__).with_name("policy.json").read_bytes()
POLICY_SHA256 = hashlib.sha256(POLICY_BYTES).hexdigest()
_POLICY = json.loads(POLICY_BYTES)

REQUEST_SCHEMA = _POLICY["request"]["schema"]
RESULT_SCHEMA = _POLICY["result_schema"]
VOCABULARY_FILE_CID = _POLICY["lexical_artifact"]["cid"]
MAX_BYTES = _POLICY["limits"]["max_bytes"]
MAX_TOKENS = _POLICY["limits"]["tokens_per_clause"]
PAD_ID = _POLICY["limits"]["padding_id"]
READER_PREFIX = tuple(_POLICY["lexical_artifact"]["reader_prefix_by_id"])
_REJECTED_IDS = frozenset(_POLICY["lexical_artifact"]["rejected_input_ids"])
TOKEN_IDS = MappingProxyType(
    {
        word: index
        for index, word in enumerate(READER_PREFIX)
        if index not in _REJECTED_IDS
    }
)
OWNERS = tuple(_POLICY["syntax"]["owners"])
OBJECTS = tuple(_POLICY["syntax"]["objects"])
LOCATIONS = tuple(_POLICY["syntax"]["locations"])
_CATEGORY_MEMBERS = {
    "O": frozenset(OWNERS),
    "D": frozenset(OWNERS),
    "X": frozenset(OBJECTS),
    "L": frozenset(LOCATIONS),
}
_FACT_FORMS = tuple(
    tuple(form.split()) for form in _POLICY["syntax"]["fact_forms"]
)
_QUERY_FORM = tuple(_POLICY["syntax"]["query_form"].split())
_PUNCTUATION = frozenset(b".,?:")
_QUERY_SUFFIX = ("?", "answer", ":")


class _Token(NamedTuple):
    word: str
    token_id: int
    start: int
    end: int


def _refuse(status: str, byte_offset: int | None) -> dict:
    return {
        "schema": RESULT_SCHEMA,
        "status": status,
        "byte_offset": byte_offset,
    }


def unavailable_artifact() -> dict:
    """Return the startup refusal without reading or parsing any request."""

    return _refuse("UNAVAILABLE_ARTIFACT", None)


def _lex(raw: bytes) -> list[_Token] | dict:
    """Scan accepted ASCII directly; every accepted byte is strict UTF-8.

    Valid non-ASCII UTF-8 and invalid UTF-8 have the same required refusal tag.
    Rejecting their first non-ASCII byte also keeps encoding/lexical failures
    in byte order rather than letting a later error replace an earlier one.
    """

    tokens: list[_Token] = []
    cursor = 0
    while cursor < len(raw):
        byte = raw[cursor]
        if byte in (32, 9, 10):
            cursor += 1
            continue
        if byte == 13:
            if cursor + 1 == len(raw) or raw[cursor + 1] != 10:
                return _refuse("INVALID_ENCODING", cursor)
            cursor += 2
            continue
        start = cursor
        if 97 <= byte <= 122:
            cursor += 1
            while cursor < len(raw) and 97 <= raw[cursor] <= 122:
                cursor += 1
            word = raw[start:cursor].decode("ascii")
            token_id = TOKEN_IDS.get(word)
            if token_id is None:
                return _refuse("UNKNOWN_LEXEME", start)
        elif byte in _PUNCTUATION:
            cursor += 1
            word = chr(byte)
            token_id = TOKEN_IDS[word]
        else:
            return _refuse("INVALID_ENCODING", cursor)
        tokens.append(_Token(word, token_id, start, cursor))
    return tokens


def _boundaries(tokens: list[_Token], end: int) -> list[list[_Token]] | dict:
    """Recover literal delimiter spans before examining admitted grammar."""

    clauses: list[list[_Token]] = []
    start = 0
    for index, token in enumerate(tokens):
        if token.word != ".":
            continue
        if index == start:
            return _refuse("UNSUPPORTED_BOUNDARY", token.start)
        if len(clauses) == 4:
            return _refuse("UNSUPPORTED_BOUNDARY", token.start)
        clauses.append(tokens[start : index + 1])
        start = index + 1
    if len(clauses) != 4 or start == len(tokens):
        return _refuse("UNSUPPORTED_BOUNDARY", end)

    query = tokens[start:]
    question_mark = next(
        (index for index, token in enumerate(query) if token.word == "?"), None
    )
    if question_mark is None:
        return _refuse("UNSUPPORTED_BOUNDARY", end)
    for offset, expected in enumerate(_QUERY_SUFFIX):
        index = question_mark + offset
        if index == len(query):
            return _refuse("UNSUPPORTED_BOUNDARY", end)
        if query[index].word != expected:
            return _refuse("UNSUPPORTED_BOUNDARY", query[index].start)
    following = question_mark + len(_QUERY_SUFFIX)
    if following != len(query):
        return _refuse("UNSUPPORTED_BOUNDARY", query[following].start)
    clauses.append(query)
    return clauses


def _matches_clause(words: tuple[str, ...], form: tuple[str, ...]) -> bool:
    """Boolean-only recognition: no captured roles or semantic assignments."""

    if len(words) != len(form):
        return False
    for word, predicate in zip(words, form, strict=True):
        members = _CATEGORY_MEMBERS.get(predicate)
        if members is None:
            if word != predicate:
                return False
        elif word not in members:
            return False
    return words[form.index("O")] != words[form.index("D")]


def _syntax_refusal(clauses: list[list[_Token]]) -> dict | None:
    words = [tuple(token.word for token in clause) for clause in clauses]
    admitted_form = next(
        (form for form in _FACT_FORMS if _matches_clause(words[0], form)), None
    )
    if admitted_form is None:
        return _refuse("UNSUPPORTED_SYNTAX", clauses[0][0].start)
    for index in range(1, 4):
        if not _matches_clause(words[index], admitted_form):
            return _refuse("UNSUPPORTED_SYNTAX", clauses[index][0].start)
    if not _matches_clause(words[4], _QUERY_FORM):
        return _refuse("UNSUPPORTED_SYNTAX", clauses[4][0].start)
    return None


def derived_input_sha256(
    inputs: Sequence[Sequence[Sequence[int]]], lengths: Sequence[Sequence[int]]
) -> str:
    """Hash the normative ordered i64 little-endian tensor framing.

    These are the transport-independent Python integer arrays of the success
    record. The caller converts them to int64 tensors only at the model seam.
    """

    if (
        len(inputs) != 1
        or len(inputs[0]) != 5
        or any(len(clause) != MAX_TOKENS for clause in inputs[0])
        or len(lengths) != 1
        or len(lengths[0]) != 5
        or any(
            type(token_id) is not int or not 0 <= token_id < 4096
            for clause in inputs[0]
            for token_id in clause
        )
        or any(
            type(length) is not int or not 1 <= length <= MAX_TOKENS
            for length in lengths[0]
        )
    ):
        raise ValueError("derived input requires i64 [1,5,13] and lengths [1,5]")
    digest = hashlib.sha256()
    for value in (
        "uor-r4.text-to-clauses-input/1",
        POLICY_SHA256,
        VOCABULARY_FILE_CID,
        "i64le",
    ):
        encoded = value.encode("utf-8")
        digest.update(struct.pack("<I", len(encoded)))
        digest.update(encoded)
    digest.update(struct.pack("<5I", 1, 5, 13, 1, 5))
    for clause in inputs[0]:
        digest.update(struct.pack("<13q", *clause))
    digest.update(struct.pack("<5q", *lengths[0]))
    return digest.hexdigest()


def segment_request(request: object) -> dict:
    """Segment one exact-schema raw request after wrapper startup verification.

    A successful result contains no role, target, semantic-world, answer or
    view fields. Refusals expose only their status and optional byte offset.
    This function has no model or artifact handles and performs zero forwards.
    """

    if (
        type(request) is not dict
        or set(request) != {"schema", "text"}
        or type(request["schema"]) is not str
        or request["schema"] != REQUEST_SCHEMA
        or type(request["text"]) is not bytes
    ):
        return _refuse("UNSUPPORTED_SCHEMA", None)
    raw = request["text"]
    if len(raw) > MAX_BYTES:
        return _refuse("INPUT_LIMIT", MAX_BYTES)

    tokens = _lex(raw)
    if isinstance(tokens, dict):
        return tokens
    clauses = _boundaries(tokens, len(raw))
    if isinstance(clauses, dict):
        return clauses
    for clause in clauses:
        if len(clause) > MAX_TOKENS:
            return _refuse("INPUT_LIMIT", clause[MAX_TOKENS].start)
    refusal = _syntax_refusal(clauses)
    if refusal is not None:
        return refusal

    inputs = [
        [
            [token.token_id for token in clause]
            + [PAD_ID] * (MAX_TOKENS - len(clause))
            for clause in clauses
        ]
    ]
    lengths = [[len(clause) for clause in clauses]]
    return {
        "schema": RESULT_SCHEMA,
        "status": "SEGMENTED",
        "policy_sha256": POLICY_SHA256,
        "raw_text_sha256": hashlib.sha256(raw).hexdigest(),
        "derived_input_sha256": derived_input_sha256(inputs, lengths),
        "clause_spans": [[clause[0].start, clause[-1].end] for clause in clauses],
        "token_spans": [
            [[token.start, token.end] for token in clause] for clause in clauses
        ],
        "inputs": inputs,
        "lengths": lengths,
    }

File: crates/uor-r4-core/src/learned_reference/adapter.rs
//! Raw-byte adapter for the frozen `R4TextToClausesV1` policy.
//!
//! The parent wrapper verifies startup artifacts and the request schema before
//! calling [`parse`]. This module reads no files and has no model handles. The
//! grammar recognizer returns only acceptance; it exposes no semantic roles.

use serde::ser::SerializeStruct;
use serde::{Serialize, Serializer};
use sha2::{Digest, Sha256};

pub(super) const REQUEST_SCHEMA: &str = "uor-r4.text-to-clauses/1";
pub(super) const RESULT_SCHEMA: &str = "uor-r4.text-to-clauses-result/1";

const POLICY: &str = include_str!(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../tools/r4-softmax-trainer/src/r4_softmax_trainer/text_clause_adapter/policy.json"
));
const VOCABULARY_FILE_CID: &str =
    "blake3:571d5fbc282b17c8726eebd7b23c3ae55212a3de81b35d27722a0fa5979b8c5b";
const MAX_BYTES: usize = 4096;
const MAX_TOKENS: usize = 13;
const PAD_ID: i64 = 57;

// Exact reader prefix in the bound policy. This is lexical metadata, not the
// core output vocabulary. IDs 0, 7, 11 and 57 are not admitted input tokens.
const READER_PREFIX: [&str; 58] = [
    "<bos>", "put", "the", "in", ".", "where", "is", "'s", "?", "answer", ":", "unknown", "mara",
    "lena", "omar", "noah", "iris", "liam", "nora", "otto", "ada", "erin", "hugo", "iona", "jude",
    "kira", "leon", "mila", "key", "book", "coin", "cup", "ring", "ball", "pen", "map", "hat",
    "toy", "jar", "box", "comb", "fork", "shoe", "doll", "drawer", "cabinet", "basket", "closet",
    "pouch", "locker", "crate", "trunk", "not", "but", ",", "owned", "by", "<pad>",
];

const FACT_FORMS: [&[&str]; 4] = [
    &[
        "O", ",", "not", "D", ",", "put", "the", "X", "in", "the", "L", ".",
    ],
    &[
        "in", "the", "L", ",", "not", "D", "but", "O", "put", "the", "X", ".",
    ],
    &[
        "not", "D", "but", "O", "put", "the", "X", "in", "the", "L", ".",
    ],
    &[
        "in", "the", "L", ",", "O", ",", "not", "D", ",", "put", "the", "X", ".",
    ],
];
const QUERY_FORM: &[&str] = &[
    "where", "is", "the", "X", "owned", "by", "O", ",", "not", "D", "?", "answer", ":",
];

/// The original refusal record, with no token, tensor or partial parse fields.
#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
pub struct Refusal {
    pub schema: &'static str,
    pub status: &'static str,
    pub byte_offset: Option<usize>,
}

impl Refusal {
    fn at(status: &'static str, byte_offset: usize) -> Self {
        Self {
            schema: RESULT_SCHEMA,
            status,
            byte_offset: Some(byte_offset),
        }
    }

    /// Refuse a malformed request before examining its raw buffer.
    pub fn unsupported_schema() -> Self {
        Self {
            schema: RESULT_SCHEMA,
            status: "UNSUPPORTED_SCHEMA",
            byte_offset: None,
        }
    }

    /// Refuse unavailable startup without examining a request.
    pub fn unavailable_artifact() -> Self {
        Self {
            schema: RESULT_SCHEMA,
            status: "UNAVAILABLE_ARTIFACT",
            byte_offset: None,
        }
    }
}

/// Internal unbatched storage; serialization preserves the original batch axis.
#[derive(Clone, Debug)]
pub(super) struct Parsed {
    pub(super) inputs: [[i64; 13]; 5],
    pub(super) lengths: [usize; 5],
    pub(super) token_spans: Vec<Vec<[usize; 2]>>,
    pub(super) clause_spans: Vec<[usize; 2]>,
    pub(super) raw_text_sha256: String,
    pub(super) derived_input_sha256: String,
}

impl Serialize for Parsed {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let mut record = serializer.serialize_struct("Parsed", 9)?;
        record.serialize_field("schema", RESULT_SCHEMA)?;
        record.serialize_field("status", "SEGMENTED")?;
        record.serialize_field("policy_sha256", &policy_sha256())?;
        record.serialize_field("raw_text_sha256", &self.raw_text_sha256)?;
        record.serialize_field("derived_input_sha256", &self.derived_input_sha256)?;
        record.serialize_field("clause_spans", &self.clause_spans)?;
        record.serialize_field("token_spans", &self.token_spans)?;
        record.serialize_field("inputs", &[&self.inputs])?;
        record.serialize_field("lengths", &[&self.lengths])?;
        record.end()
    }
}

#[derive(Clone, Copy)]
struct Token<'a> {
    word: &'a [u8],
    id: i64,
    start: usize,
    end: usize,
}

pub(super) fn policy_sha256() -> String {
    hex::encode(Sha256::digest(POLICY.as_bytes()))
}

fn lex(raw: &[u8]) -> Result<Vec<Token<'_>>, Refusal> {
    let mut tokens = Vec::new();
    let mut cursor = 0;
    while cursor < raw.len() {
        let byte = raw[cursor];
        match byte {
            b' ' | b'\t' | b'\n' => {
                cursor += 1;
                continue;
            }
            b'\r' => {
                if raw.get(cursor + 1) != Some(&b'\n') {
                    return Err(Refusal::at("INVALID_ENCODING", cursor));
                }
                cursor += 2;
                continue;
            }
            _ => {}
        }

        let start = cursor;
        let id = match byte {
            b'a'..=b'z' => {
                cursor += 1;
                while cursor < raw.len() && raw[cursor].is_ascii_lowercase() {
                    cursor += 1;
                }
                let word = &raw[start..cursor];
                let found = READER_PREFIX.iter().enumerate().find(|(index, value)| {
                    !matches!(*index, 0 | 7 | 11 | 57) && value.as_bytes() == word
                });
                match found {
                    Some((index, _)) => index as i64,
                    None => return Err(Refusal::at("UNKNOWN_LEXEME", start)),
                }
            }
            b'.' => {
                cursor += 1;
                4
            }
            b',' => {
                cursor += 1;
                54
            }
            b'?' => {
                cursor += 1;
                8
            }
            b':' => {
                cursor += 1;
                10
            }
            _ => return Err(Refusal::at("INVALID_ENCODING", cursor)),
        };
        tokens.push(Token {
            word: &raw[start..cursor],
            id,
            start,
            end: cursor,
        });
    }
    Ok(tokens)
}

fn boundaries<'a, 'raw>(
    tokens: &'a [Token<'raw>],
    end: usize,
) -> Result<Vec<&'a [Token<'raw>]>, Refusal> {
    let mut clauses = Vec::with_capacity(5);
    let mut start = 0;
    for (index, token) in tokens.iter().enumerate() {
        if token.word != b"." {
            continue;
        }
        if index == start || clauses.len() == 4 {
            return Err(Refusal::at("UNSUPPORTED_BOUNDARY", token.start));
        }
        clauses.push(&tokens[start..=index]);
        start = index + 1;
    }
    if clauses.len() != 4 || start == tokens.len() {
        return Err(Refusal::at("UNSUPPORTED_BOUNDARY", end));
    }

    let query = &tokens[start..];
    let Some(question_mark) = query.iter().position(|token| token.word == b"?") else {
        return Err(Refusal::at("UNSUPPORTED_BOUNDARY", end));
    };
    for (offset, expected) in [b"?".as_slice(), b"answer".as_slice(), b":".as_slice()]
        .into_iter()
        .enumerate()
    {
        let Some(token) = query.get(question_mark + offset) else {
            return Err(Refusal::at("UNSUPPORTED_BOUNDARY", end));
        };
        if token.word != expected {
            return Err(Refusal::at("UNSUPPORTED_BOUNDARY", token.start));
        }
    }
    if let Some(token) = query.get(question_mark + 3) {
        return Err(Refusal::at("UNSUPPORTED_BOUNDARY", token.start));
    }
    clauses.push(query);
    Ok(clauses)
}

fn matches_clause(clause: &[Token<'_>], form: &[&str]) -> bool {
    if clause.len() != form.len() {
        return false;
    }
    for (token, predicate) in clause.iter().zip(form) {
        let accepted = match *predicate {
            "O" | "D" => (12..28).contains(&token.id),
            "X" => (28..44).contains(&token.id),
            "L" => (44..52).contains(&token.id),
            literal => token.word == literal.as_bytes(),
        };
        if !accepted {
            return false;
        }
    }
    // The two membership predicates must name distinct owner words. This is a
    // boolean syntax condition; no positions or assignments leave this function.
    let owner = form.iter().position(|predicate| *predicate == "O");
    let distractor = form.iter().position(|predicate| *predicate == "D");
    match (
        owner.and_then(|index| clause.get(index)),
        distractor.and_then(|index| clause.get(index)),
    ) {
        (Some(owner), Some(distractor)) => owner.word != distractor.word,
        _ => false,
    }
}

fn syntax(clauses: &[&[Token<'_>]]) -> Result<(), Refusal> {
    // `boundaries` is the only caller's producer and supplies five nonempty
    // clauses. Indexing here does not depend on a caller-provided clause array.
    let Some(form) = FACT_FORMS
        .iter()
        .find(|form| matches_clause(clauses[0], form))
    else {
        return Err(Refusal::at("UNSUPPORTED_SYNTAX", clauses[0][0].start));
    };
    for clause in &clauses[1..4] {
        if !matches_clause(clause, form) {
            return Err(Refusal::at("UNSUPPORTED_SYNTAX", clause[0].start));
        }
    }
    if !matches_clause(clauses[4], QUERY_FORM) {
        return Err(Refusal::at("UNSUPPORTED_SYNTAX", clauses[4][0].start));
    }
    Ok(())
}

fn derived_input_sha256(inputs: &[[i64; 13]; 5], lengths: &[usize; 5]) -> String {
    let mut digest = Sha256::new();
    let policy_sha256 = policy_sha256();
    for value in [
        "uor-r4.text-to-clauses-input/1",
        policy_sha256.as_str(),
        VOCABULARY_FILE_CID,
        "i64le",
    ] {
        digest.update((value.len() as u32).to_le_bytes());
        digest.update(value.as_bytes());
    }
    for dimension in [1u32, 5, 13, 1, 5] {
        digest.update(dimension.to_le_bytes());
    }
    for clause in inputs {
        for id in clause {
            digest.update(id.to_le_bytes());
        }
    }
    for length in lengths {
        digest.update((*length as i64).to_le_bytes());
    }
    hex::encode(digest.finalize())
}

/// Parse raw bytes after the wrapper has accepted startup and request schema.
pub(super) fn parse(raw: &[u8]) -> Result<Parsed, Refusal> {
    if raw.len() > MAX_BYTES {
        return Err(Refusal::at("INPUT_LIMIT", MAX_BYTES));
    }
    let tokens = lex(raw)?;
    let clauses = boundaries(&tokens, raw.len())?;
    for clause in &clauses {
        if let Some(token) = clause.get(MAX_TOKENS) {
            return Err(Refusal::at("INPUT_LIMIT", token.start));
        }
    }
    syntax(&clauses)?;

    let mut inputs = [[PAD_ID; 13]; 5];
    let mut lengths = [0; 5];
    let mut token_spans = Vec::with_capacity(5);
    let mut clause_spans = Vec::with_capacity(5);
    for (index, clause) in clauses.iter().enumerate() {
        lengths[index] = clause.len();
        let mut spans = Vec::with_capacity(clause.len());
        for (position, token) in clause.iter().enumerate() {
            inputs[index][position] = token.id;
            spans.push([token.start, token.end]);
        }
        token_spans.push(spans);
        clause_spans.push([clause[0].start, clause[clause.len() - 1].end]);
    }
    Ok(Parsed {
        raw_text_sha256: hex::encode(Sha256::digest(raw)),
        derived_input_sha256: derived_input_sha256(&inputs, &lengths),
        inputs,
        lengths,
        token_spans,
        clause_spans,
    })
}
